### PR TITLE
Fix bugs of addLocalData() and add tests for edge cases

### PR DESCRIPTION
**Pull request by pochenai**
- Add tests for attacking the mid-right and right most branch and fix related bug

### CODE REVIEW

File: packages/contracts-bedrock/src/dispute/FaultDisputeGameN.sol [hunk 1]
```diff
@@ -517,7 +517,12 @@ contract FaultDisputeGame is IFaultDisputeGame, Clone, ISemver {
             value_ = starting.raw();
         } else if (_ident == LocalPreimageKey.DISPUTED_OUTPUT_ROOT) {
             // Load the disputed proposal's output root
-            Claim disputed = getClaim(disputedRoot.raw(), disputedPos, _daItem);
+            Claim disputed;
+            if (disputedPos.raw() == 1) {
```
Comment (qizhou) [inline packages/contracts-bedrock/src/dispute/FaultDisputeGameN.sol:522]: I would suggest adding a comment here

File: packages/contracts-bedrock/src/dispute/FaultDisputeGameN.sol [hunk 2]
```diff
@@ -1115,7 +1120,7 @@ contract FaultDisputeGame is IFaultDisputeGame, Clone, ISemver {
         if (ancestorPos_.depth() == SPLIT_DEPTH + N_BITS) {
             ancestorClaim_ = ancestorClaimRoot;
         } else {
-            ancestorClaim_ = getClaim(ancestorClaimRoot.raw(), _pos, _daItem);
+            ancestorClaim_ = getClaim(ancestorClaimRoot.raw(), Position.wrap(_pos.raw() - 1), _daItem);
```
Comment (qizhou) [inline packages/contracts-bedrock/src/dispute/FaultDisputeGameN.sol]: Should `_pos` be replaced by `ancestorPos_`?

File: packages/contracts-bedrock/src/dispute/FaultDisputeGameN.sol [hunk 3]
```diff
@@ -1156,10 +1161,14 @@ contract FaultDisputeGame is IFaultDisputeGame, Clone, ISemver {
         // Walk up the DAG to find a claim that commits to the same trace index as `_pos`. It is
         // guaranteed that such a claim exists.
         ClaimData storage ancestor_ = claimData[_start];
-        while (ancestor_.position.raw() != traceAncestorPosValue) {
-            ancestor_ = claimData[ancestor_.parentIndex];
+        if (traceAncestorPosValue == 0) {
```
Comment (qizhou) [inline packages/contracts-bedrock/src/dispute/FaultDisputeGameN.sol]: If `traceAncestorPosValue` should the claim be `starting` one?  Or in what situation that `traceAncestorPosValue` is zero?

File: packages/contracts-bedrock/src/dispute/FaultDisputeGameN.sol (hunk 1 above)
Comment (dajuguan) [reply packages/contracts-bedrock/src/dispute/FaultDisputeGameN.sol:522]: > I would suggest adding a comment here

Fixed here

File: packages/contracts-bedrock/src/dispute/FaultDisputeGameN.sol [hunk 4]
```diff
@@ -518,8 +518,9 @@ contract FaultDisputeGame is IFaultDisputeGame, Clone, ISemver {
         } else if (_ident == LocalPreimageKey.DISPUTED_OUTPUT_ROOT) {
             // Load the disputed proposal's output root
             Claim disputed;
+            // If the pos is 1, then the rootclaim itself is the output hash.
             if (disputedPos.raw() == 1) {
```
Comment (pochenai) [inline packages/contracts-bedrock/src/dispute/FaultDisputeGameN.sol:522]: Fix https://github.com/ethstorage/optimism/pull/41/commits/74242dce1d3fa38ca34c3629e0094a1ee292727a#r1683739419

File: packages/contracts-bedrock/src/dispute/FaultDisputeGameN.sol (hunk 1 above)
Comment (pochenai) [reply packages/contracts-bedrock/src/dispute/FaultDisputeGameN.sol:522]: Fixed in https://github.com/ethstorage/optimism/pull/41/commits/74242dce1d3fa38ca34c3629e0094a1ee292727a#r1683759252

File: packages/contracts-bedrock/src/dispute/FaultDisputeGameN.sol (hunk 2 above)
Comment (pochenai) [reply packages/contracts-bedrock/src/dispute/FaultDisputeGameN.sol]: Actually, `ancestorPos_` should be the right context, but the `ancestorPos_` value is not right. Maybe we should rename `_pos` to `ancestorPos_`?

File: packages/contracts-bedrock/src/dispute/FaultDisputeGameN.sol [hunk 5]
```diff
@@ -1154,14 +1155,15 @@ contract FaultDisputeGame is IFaultDisputeGame, Clone, ISemver {
             : _firstValidRightIndex(_pos.traceAncestorBounded(SPLIT_DEPTH), N_BITS);
 
         uint256 offset = ancestorPos_.raw() % (1 << N_BITS);
-        if (MAX_ATTACK_BRANCH == offset) {
+        // If ancestorPos_.raw() == 1 (rootclaim), we don't offset it
+        if (MAX_ATTACK_BRANCH == offset || ancestorPos_.raw() == 1) {
```
Comment (pochenai) [inline packages/contracts-bedrock/src/dispute/FaultDisputeGameN.sol]: Let offset be 0 when its rootClaim, so the meaning of `traceAncestorPosValue ` will be consistent.

File: packages/contracts-bedrock/src/dispute/FaultDisputeGameN.sol (hunk 3 above)
Comment (pochenai) [reply packages/contracts-bedrock/src/dispute/FaultDisputeGameN.sol]: Because we offset it by 1, so it's 0 here. For consistency, I've made changes to the offset logic and let the traceAncestorPosValue be 1. https://github.com/ethstorage/optimism/pull/41/commits/74242dce1d3fa38ca34c3629e0094a1ee292727a#r1683762415

File: packages/contracts-bedrock/src/dispute/FaultDisputeGameN.sol [hunk 6]
```diff
@@ -1149,8 +1155,9 @@ contract FaultDisputeGame is IFaultDisputeGame, Clone, ISemver {
             : _firstValidRightIndex(_pos.traceAncestorBounded(SPLIT_DEPTH), N_BITS);
 
         uint256 offset = ancestorPos_.raw() % (1 << N_BITS);
-        if (MAX_ATTACK_BRANCH == offset) {
-            offset = 0;
+        // If ancestorPos_.raw() == 1 (rootclaim), we don't offset it
```
Comment (qizhou) [inline packages/contracts-bedrock/src/dispute/FaultDisputeGameN.sol]: Comment should be updated accordingly.

**Review by qizhou — APPROVED** [review]

File: packages/contracts-bedrock/src/dispute/FaultDisputeGameN.sol (hunk 6 above)
Comment (pochenai) [reply packages/contracts-bedrock/src/dispute/FaultDisputeGameN.sol]: Fixed
